### PR TITLE
[jest] add `jest.Spied*` utility types

### PR DESCRIPTION
**Pull request by mrazauskas**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/facebook/jest/pull/13440

### Summary

Jest [v29.2.1](https://github.com/facebook/jest/releases/tag/v29.2.1) just got released with `jest.Spied*` utility types, which were added in https://github.com/facebook/jest/pull/13440. I adapted typings from this PR and added tests.

### CODE REVIEW

File: types/jest/index.d.ts [hunk 1]
```diff
@@ -26,6 +26,7 @@
 //                 Pawel Fajfer <https://github.com/pawfa>
 //                 Alexandre Germain <https://github.com/gerkindev>
 //                 Adam Jones <https://github.com/domdomegg>
+//                 Tom Mrazauskas <https://github.com/mrazauskas>
```
Comment (mrazauskas) [inline types/jest/index.d.ts:29]: It would be nice to be notified if types are changing here, so I could adjust types in Jest repo.

**Review by domdomegg — APPROVED** [review]

**Review by sandersn — APPROVED** [review]

**Comment by mrazauskas** [comment]
Ready to merge

File: types/jest/index.d.ts [hunk 2]
```diff
@@ -389,18 +389,18 @@ declare namespace jest {
...
-    ): FunctionProperties<Required<T>>[M] extends Func
-        ? SpyInstance<ReturnType<FunctionProperties<Required<T>>[M]>, ArgsType<FunctionProperties<Required<T>>[M]>>
-        : never;
     function spyOn<T extends {}, M extends ConstructorPropertyNames<Required<T>>>(
         object: T,
         method: M,
     ): Required<T>[M] extends new (...args: any[]) => any
         ? SpyInstance<InstanceType<Required<T>[M]>, ConstructorArgsType<Required<T>[M]>>
         : never;
+    function spyOn<T extends {}, M extends FunctionPropertyNames<Required<T>>>(
```
Comment (dlech) [inline types/jest/index.d.ts:399]: Swapping the order here breaks this use case:

```typescript
jest.spyOn(Storage.prototype, 'getItem').mockImplementation((key) => null);
```

undoing this change breaks

```typescript
let typedSpy1: jest.SpiedClass<typeof globalThis.Date>;
typedSpy1 = jest.spyOn(globalThis, 'Date');
```

How can we make both work?

Comment (mrazauskas) [reply types/jest/index.d.ts:399]: See #63182. Seems like it will make both cases work as expected.

Comment (dlech) [reply types/jest/index.d.ts:399]: Opened #63196 to address this.

Comment (dlech) [reply types/jest/index.d.ts:399]: Missed that one, thanks!

Comment (mrazauskas) [reply types/jest/index.d.ts:399]: Just to add. As you noticed swapping the order made `spyOn` work correctly with `Date`. The type of `Date` is rather odd indeed, but the same could be said about `Storage`. Not so easy to cover both.

**Pull request merged**